ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java

```diff
@@ -765,6 +768,32 @@ private static void start(SessionState startSs, boolean isAsync, LogHelper conso
     }
   }
 
+  /**
+   * Set hive metastore delegation token in current UGI
+   */
+  private void setHiveMetastoreDelegationTokenToUGI() {
+    if (this.isHiveServerQuery) {
+      return;
+    }
+    if (!getSessionConf().getBoolVar(ConfVars.METASTORE_USE_THRIFT_SASL)) {
+      return;
+    }
+    try {
+      UserGroupInformation UGI = Utils.getUGI();
+      String hmsDelegationTokenStr = Hive.get().getDelegationToken(UGI.getShortUserName(), UGI.getShortUserName());
```

Comment:
I think `Hive.get().getDelegationToken` will create a connection to metastore, and the session can reuse this Hive client later on. It looks the token got in this step takes no effect.

Reply:
> I think `Hive.get().getDelegationToken` will create a connection to metastore, and the session can reuse this Hive client later on. It looks the token got in this step takes no effect.

For iceberg issue https://github.com/apache/iceberg/issues/3127,insert into iceberg table with hive client ,iceberg table commit action occurs during the ApplicationMaster commit job phase without kerberos info,So we can only rely on HiveMetastore DelegationToken to solve this problem.

"will create a connection to metastore" means new HiveMetastoreClient , However, the hive client does not enable the "HADOOP_PROXY_USER", which means that no DelegationToken will be stuffed into the UGI.

Reply:
If this new method is called from the Kerberos yarn client(which is kerberos based), and want to pass the HMS token to yarn job, some thoughts:
1. what happen if the token being expired, should we get a new one? 
2. will the token in session config be propagated to the yarn job?

and I think we also need to check the client is kerberos based before fetching the dt: `UserGroupInformation.isSecurityEnabled()`
As this is Iceberg specific, an alternative way is to move this part to `configureJobConf`:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java#L198

Reply:
> If this new method is called from the Kerberos yarn client(which is kerberos based), and want to pass the HMS token to yarn job, some thoughts:
> 
> 1. what happen if the token being expired, should we get a new one?
> 2. will the token in session config be propagated to the yarn job?
> 
> and I think we also need to check the client is kerberos based before fetching the dt: `UserGroupInformation.isSecurityEnabled()` As this is Iceberg specific, an alternative way is to move this part to `configureJobConf`: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java#L198

1.if the token being expired , we can  do renew token such as  https://github.com/apache/hive/blob/698194183a2cb4ee6bc528ee71462cbae2c2c7ad/service/src/java/org/apache/hive/service/cli/session/HiveSessionImplwithUGI.java#L142-L161 .
i am not sure should we do this.
2.if token has set in UGI,new HMS client will check ConfVars.TOKEN_SIGNATURE for tokenSig.which ConfVars.TOKEN_SIGNATURE  has set in SessionState.so that we can connect to HMS (which is kerberos based) with HMS token

Reply:
> `hive.metastore.sasl.enabled = true` means HiveMetastore thrift interface will be secured with SASL. Clients must authenticate with Kerberos. so `UserGroupInformation.isSecurityEnabled()` is not necessary

Reply:
Hi @hzxiongyinke,
> 1.if the token being expired , we can do renew token such as

I'm not sure how to trigger the token renew in HiveSessionImplWithUGI, as I understand from the context, this happens on Hive client, not HiveServer2.

> if token has set in UGI,new HMS client will check ConfVars.TOKEN_SIGNATURE for tokenSig.which ConfVars.TOKEN_SIGNATURE has set in SessionState

That's right, what I concern is that the App master doesn't sit inside the Hive client, they are two different processes, the App master may donot know the token in Hive client.

> Clients must authenticate with Kerberos

Right, this is a property for connecting to HMS, it doesn't mean the client itself is Kerberos based, I would prefer add the check here in my opinion.